### PR TITLE
State migration spike

### PR DESCRIPTION
**Pull request by LesnyRumcajs**
## Summary of changes

Changes introduced in this pull request:
- added state migration spike results to the developer documentation.

## Reference issue to close (if applicable)

Closes https://github.com/ChainSafe/forest/issues/2657

## Change checklist

- [x] I have performed a self-review of my own code,
- [x] I have made corresponding changes to the documentation,
- [x] I have added tests that prove my fix is effective or that my feature works (if possible),
- [x] I have made sure the [CHANGELOG](https://github.com/ChainSafe/forest/blob/main/CHANGELOG.md) is up-to-date. All user-facing changes should be reflected in this document.

### CODE REVIEW

File: documentation/src/developer_documentation/state_migration_spike.md [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+Iterate over the state of each actor at the given epoch and write the new state along with any specific changes to the respective state. This involves iterating over each of the HAMT nodes storing the state and writing them to the database.
+
+[Lotus upgrade method](https://github.com/filecoin-project/lotus/blob/58900a70333a11a903cf9fe3f29e6a5c309cb000/chain/consensus/filcns/upgrades.go#L1591-L1612) and the [module](https://github.com/filecoin-project/go-state-types/tree/master/builtin/v10/migration) dedicated to Actors `v10` migration. The core logic is [here](https://github.com/filecoin-project/go-state-types/blob/master/builtin/v10/migration/top.go#L28). The same module is used by Venus.
+
+Forks migrations: handled by [fork.go](https://github.com/filecoin-project/lotus/blob/58900a70333a11a903cf9fe3f29e6a5c309cb000/chain/stmgr/forks.go#L42-L53) entities.
+
+### Where to upgrade?
+
+It should be done ost likely in the apply blocks method.
```
Comment (lemmih) [inline documentation/src/developer_documentation/state_migration_spike.md:44]: `most`?

```suggestion
It should be done most likely in the apply blocks method.
```

**Review by lemmih — APPROVED** [review]
Good overview, good plan.

File: documentation/src/developer_documentation/state_migration_spike.md [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+State migration is a process where the `StateTree` contents are transformed from
+an older form to a newer form.
+
+### Why do we need to migrate?
+
+Migration is required when the structure of the state changes. This happens when
+new fields are added or existing ones are modified. Migration is **not**
+required in case of new behaviour.
+
+In case of NV18, the `StateTree` changed from version 4 to version 5. See
```
Comment (ec2) [inline documentation/src/developer_documentation/state_migration_spike.md:14]: For the example of NV18, there's more than just the StateTree version changing. For example, the init actor is being [migrated](https://github.com/filecoin-project/go-state-types/blob/master/builtin/v10/migration/init.go#L32) as well

File: documentation/src/developer_documentation/state_migration_spike.md [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+https://github.com/filecoin-project/ref-fvm/pull/1062
+
+### What to upgrade?
+
+We need to upgrade the `StateTree` which is represented as
```
Comment (ec2) [inline documentation/src/developer_documentation/state_migration_spike.md:19]: Same as comment above. It's more than just upgrading the StateTree for NV18 it seems.

**Review by elmattic — APPROVED** [review]
Looking good to me.

I'm wondering about interaction between our new GC code and this migration framework but also if we encounter edge case scenarios during the migration (the user stops the node, or Forest is stopping with a recoverable error, some lost connectivity issue, etc).

For testing purposes it would definitely help to have some CI script downloading automatically a pre upgrade snapshot for calibnet, in case our local devnet upgrade works but the calibnet one doesn't.

**Pull request merged**